### PR TITLE
Undeploy

### PR DESCRIPTION
**Pull request by flurdy**
Undeploy service and endpoint.

Includes temporary toggles until PBE endpoint for registering unemployment is done, 
and for workflow support of removing deployments

### CODE REVIEW

File: src/api/undeploy/controllers/index.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { undeployServiceController } from '~/src/api/undeploy/controllers/undeploy-service.js'
+
+export { undeployServiceController }
```
Comment (philsegal) [inline src/api/undeploy/controllers/index.js:3]: Is this needed?

**Review by philsegal — APPROVED** [review]

File: src/api/undeploy/controllers/index.js (hunk 1 above)
Comment (flurdy) [reply src/api/undeploy/controllers/index.js:3]: Meh, Can be explicit in the router to the full controller path however once we add another controller we have to change it all again...

Comment (feedmypixel) [reply src/api/undeploy/controllers/index.js:3]: I've been really trying to cut down on excessive `index.js` files. Same same, i'd say this isn't needed

File: src/api/undeploy/helpers/github/delete-github-file.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  commitMessage,
+  filePath,
+  logger
+) {
+  let startTime = Date.now()
+  const commitSha = await getLatestCommitSha(owner, repo, branch)
+  logger.info(`getLatestCommitSha: ${Date.now() - startTime}ms`)
+
+  startTime = Date.now()
+
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/github/delete-github-file.js:26]: with `graphQL` in the codebase a few helpers would be of help:
- https://www.npmjs.com/package/eslint-plugin-graphql
- https://www.npmjs.com/package/graphql-tag

File: src/api/undeploy/helpers/remove-deployment-file.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * @param {{id: string, displayName: string}} user
+ * @param {Logger} logger
+ */
+async function removeDeploymentFile(
+  undeploymentId,
+  imageName,
+  environment,
+  zone,
+  user,
+  logger
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/remove-deployment-file.js:22]: how come logger is being passed along? Can you not just used `createLogger()`?

File: src/api/undeploy/helpers/schema/undeploy-service-validation.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import Joi from 'joi'
+
+import { environments } from '~/src/config/index.js'
+
+function undeployServiceValidation() {
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/schema/undeploy-service-validation.js:5]: generally if this is an object then just export the object. If it needs values export a function

File: src/api/undeploy/helpers/undeploy-service-from-environment.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * @param {string} imageName
+ * @param {string} environment
+ * @param {{id: string, displayName: string}} user
+ * @param {Logger} logger
+ */
+async function undeployServiceFromEnvironment(
+  imageName,
+  environment,
+  user,
+  logger
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/undeploy-service-from-environment.js:20]: `logger` appears to be going on some sort of long journey. Is this needed? can you just use `createLogger()` in the final destination?

File: src/api/undeploy/helpers/undeploy-service-from-environment.test.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  return {
+    registerUndeployment: jest.fn()
+  }
+})
+jest.mock('~/src/api/undeploy/helpers/remove-deployment-file', () => {
+  return {
+    removeDeploymentFile: jest.fn()
+  }
+})
+jest.mock('~/src/api/deploy/helpers/lookup-tenant-service', () => {
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/undeploy-service-from-environment.test.js:27]: mocking is a bit of a smell here. Is all this mocking needed?

File: src/api/undeploy/helpers/undeploy-service-from-environment.test.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  return {
+    isFeatureEnabled: jest.fn()
+  }
+})
+jest.mock('~/src/api/undeploy/helpers/register-undeployment', () => {
+  return {
+    registerUndeployment: jest.fn()
+  }
+})
+jest.mock('~/src/api/undeploy/helpers/remove-deployment-file', () => {
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/undeploy-service-from-environment.test.js:22]: I'm pretty sure all of these can be `jest.mock('~/src/api/undeploy/helpers/remove-deployment-file')` which mocks all items in the file

File: src/api/undeploy/helpers/undeploy-service-from-environment.test.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    imageName,
+    environment,
+    user,
+    logger
+  )
+}
+
+describe('#undeployServiceFromEnvironment', () => {
+  beforeEach(() => {
+    jest.clearAllMocks()
```
Comment (feedmypixel) [inline src/api/undeploy/helpers/undeploy-service-from-environment.test.js:54]: this is being done globally

File: src/api/undeploy/helpers/schema/undeploy-service-validation.js (hunk 4 above)
Comment (flurdy) [reply src/api/undeploy/helpers/schema/undeploy-service-validation.js:5]: could do. I just copied the deploy validation and trimmed it

File: src/api/undeploy/helpers/undeploy-service-from-environment.test.js (hunk 7 above)
Comment (flurdy) [reply src/api/undeploy/helpers/undeploy-service-from-environment.test.js:22]: :confused:

File: src/api/undeploy/helpers/schema/undeploy-service-validation.js (hunk 4 above)
Comment (feedmypixel) [reply src/api/undeploy/helpers/schema/undeploy-service-validation.js:5]: yeah its not a showstopper, more preference

File: src/api/undeploy/helpers/remove-deployment-file.js (hunk 3 above)
Comment (flurdy) [reply src/api/undeploy/helpers/remove-deployment-file.js:22]: Yes though I thought we'd want to keep request id etc?

File: src/api/undeploy/helpers/undeploy-service-from-environment.test.js (hunk 7 above)
Comment (flurdy) [reply src/api/undeploy/helpers/undeploy-service-from-environment.test.js:22]: I am not sure I get this one

**Pull request merged**